### PR TITLE
feat: support cache all at streaming

### PR DESCRIPTION
**Pull request by cutecutecat**
# Revision
- Add argument `forceCacheAllEvenLazy` to force cache all files at GCP image streaming
- Split functions `commonJob`, `dirCacheJob` and `fullCacheJob` for different cache method

| method       | discription                                  |
| ------------ | -------------------------------------------- |
| commonJob    | cache nothing at streaming mode              |
| dirCacheJob  | cache selected directories at streaming mode |
| fullCacheJob | cache all files at streaming mode            |

When `forceCacheAllEvenLazy` is enabled, `kube-fledged` would always use `fullCacheJob`.
【Forward compatibility】Otherwise, `kube-fledged` would use `dirCacheJob` for `modelzai` images and `commonJob` for others.

### CODE REVIEW

**Review by gaocegege — COMMENTED** [review]
Thanks for your contribution! :tada: :+1:

Have you tested the full cache job? Will it take too long to finish?

**Comment by cutecutecat** [comment]
> Thanks for your contribution! 🎉 👍
> 
> Have you tested the full cache job? Will it take too long to finish?

Sorry about that, I tested it recently, with the image `nvidia/cuda:11.7.1-base-ubuntu22.04` which is the base layer of most modelz templates.

Running `time find / -type f -print0 | xargs -0 cat > /dev/null || true` would cost `>40` minutes to finish, and it hasn't finished yet.

```shell
> docker run --rm -it nvidia/cuda:11.7.1-base-ubuntu22.04 bash
root@a6c516b9bc01:/# time find / -type f -print0 | xargs -0 cat > /dev/null || true
^C

real    41m0.038s
user    0m0.020s
sys     0m0.009s
```

I think this might be unacceptable, maybe we would turn to survey about common directories used by modern frameworks? 

Update:

It seems that `cat` at huge files would slow/block the process, remove that would end swiftly. I would try to find some `cat` alternatives.

Update: 
After exclude 3 directories, everything is fine.
`find / -not -path "/dev/*" -not -path "/proc/*" -not -path "/sys/*"  -type f -print0 | xargs -0 cat > /dev/null || true`.
```shell
root@a6c516b9bc01:/# time find / -not -path "/dev/*" -not -path "/proc/*" -not -path "/sys/*"  -type f -print0 | xargs -0 cat > /dev/null || true

real    0m0.085s
user    0m0.031s
sys     0m0.057s
```
I would update this command soon.

File: pkg/apis/kubefledged/v1alpha2/types.go [hunk 1]
```diff
@@ -38,8 +38,9 @@ type ImageCache struct {
 
 // CacheSpecImages specifies the Images to be cached
 type CacheSpecImages struct {
-	Images       []string          `json:"images"`
-	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+	Images        []string          `json:"images"`
```
Comment (xieydd) [inline pkg/apis/kubefledged/v1alpha2/types.go]: If you create a structure, will it be modified much?
For example:
```
type Image struct {
    Name string
    ForceCacheAll bool
}
```

Comment (xieydd) [reply pkg/apis/kubefledged/v1alpha2/types.go]: It looks like you need change the crd version if you use struct.
It may be a little bit complex.

**Review by xieydd — COMMENTED** [review]
some nits

File: pkg/apis/kubefledged/v1alpha2/types.go (hunk 1 above)
Comment (gaocegege) [reply pkg/apis/kubefledged/v1alpha2/types.go]: Yeah. If the existing fields' struct is changed, Kubernetes requires a version upgrade.

Comment (cutecutecat) [reply pkg/apis/kubefledged/v1alpha2/types.go]: Moved to a new version `v1alpha3`

> If you create a structure, will it be modified much? For example:
> 
> ```
> type Image struct {
>     Name string
>     ForceCacheAll bool
> }
> ```

Created a new version of `alpha` with `Image` type.

File: cmd/migration/main.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	flag.Parse()
+
+	clientCmdConfig, err := clientcmd.BuildConfigFromFlags(masterURL, kubeConfig)
+	if err != nil {
+		glog.Fatalf("error building kubeconfig: %s", err.Error())
+	}
+
+	client, err := clientset.NewForConfig(clientCmdConfig)
+	if err != nil {
+		glog.Fatalf("error building Inference clientset: %s", err.Error())
```
Comment (xieydd) [inline cmd/migration/main.go:41]: Inference -> kubeFledged

**Review by xieydd — APPROVED** [review]
some nits.
Other LGTM.

File: cmd/controller/main.go [hunk 3]
```diff
@@ -78,7 +78,7 @@ func main() {
 
 	controller := app.NewController(kubeClient, fledgedClient, fledgedNameSpace,
 		kubeInformerFactory.Core().V1().Nodes(),
-		fledgedInformerFactory.Kubefledged().V1alpha2().ImageCaches(),
+		fledgedInformerFactory.Kubefledged().V1alpha3().ImageCaches(),
```
Comment (gaocegege) [inline cmd/controller/main.go:81]: Do we need upgrade version?

**Comment by gaocegege** [comment]
lgtm

**Pull request merged**